# Spotify Analysis

In [1]:
import pandas as pd
import os

In [3]:
df=pd.read_csv("./spotify_data.csv")

In [4]:
df

playlist      song_name         singer           album  \
0   Today’s Top Hits          Water           Tyla           Water   
1   Today’s Top Hits    Lovin On Me    Jack Harlow     Lovin On Me   
2   Today’s Top Hits         greedy     Tate McRae          greedy   
3   Today’s Top Hits        Houdini       Dua Lipa         Houdini   
4   Today’s Top Hits       MY HOUSE        Beyoncé        MY HOUSE   
..               ...            ...            ...             ...   
93             Sleep  The New World     Cardi Mist  Need Alone TIm   
94             Sleep      Following     Aquilonius       Following   
95             Sleep  Herinneringen     Sohn Aelia   Herinneringen   
96             Sleep       Scorchio  Maria Botican    Infinity Air   
97             Sleep      Boundless         Haldur       Boundless   

   song_duration  
0           3:20  
1           2:18  
2           2:11  
3           3:05  
4           4:22  
..           ...  
93          3:14  
94          3:05  
95          2:52  
96          3:16  
97          2:58  

[98 rows x 5 columns]

In [5]:
# I don't think that cleaning of data is required but if required we do in questions only

## Q1.What is the average duration of songs in each playlist, and which playlist has the highest average song duration?

In [10]:
df1=df

In [17]:
df['song_duration'].isna().sum()

0

In [18]:
# Split the 'song_duration' column based on ':' to get the time in seconds
df['song_duration_insec'] = (df['song_duration'].str.split(':').str[0].astype(int))*60 + df['song_duration'].str.split(':').str[1].astype(int)

In [19]:
df

playlist      song_name         singer           album  \
0   Today’s Top Hits          Water           Tyla           Water   
1   Today’s Top Hits    Lovin On Me    Jack Harlow     Lovin On Me   
2   Today’s Top Hits         greedy     Tate McRae          greedy   
3   Today’s Top Hits        Houdini       Dua Lipa         Houdini   
4   Today’s Top Hits       MY HOUSE        Beyoncé        MY HOUSE   
..               ...            ...            ...             ...   
93             Sleep  The New World     Cardi Mist  Need Alone TIm   
94             Sleep      Following     Aquilonius       Following   
95             Sleep  Herinneringen     Sohn Aelia   Herinneringen   
96             Sleep       Scorchio  Maria Botican    Infinity Air   
97             Sleep      Boundless         Haldur       Boundless   

   song_duration  song_duration_insec  
0           3:20                  200  
1           2:18                  138  
2           2:11                  131  
3           3:05                  185  
4           4:22                  262  
..           ...                  ...  
93          3:14                  194  
94          3:05                  185  
95          2:52                  172  
96          3:16                  196  
97          2:58                  178  

[98 rows x 6 columns]

In [20]:
grouped_df = df.groupby('playlist').agg({'song_duration_insec': 'mean'})

In [21]:
grouped_df

song_duration_insec
playlist                             
Calming Acoustic           163.083333
Christmas Hits             183.480000
Sleep                      183.125000
Today’s Top Hits           201.200000

In [41]:
df['song_duration_insec'].mean()

182.91836734693877

In [42]:
Calming_df=df[(df['playlist'] == "Calming Acoustic") & (df['song_duration_insec'] >= 175) & (df['song_duration_insec'] <= 190)]

In [43]:
Calming_df

playlist          song_name            singer             album  \
59  Calming Acoustic    Act of Kindness       Amma Lerato   Act of Kindness   
62  Calming Acoustic  Blank Stare Happy   American Legion      Seeking Song   
69  Calming Acoustic   Lascia Fare A Me  Valerio Castilla  Lascia Fare A Me   
71  Calming Acoustic           My Place      Denver Paige          My Place   

   song_duration  song_duration_insec  
59          3:05                  185  
62          2:56                  176  
69          3:05                  185  
71          3:07                  187

In [44]:
len(Calming_df)

4

In [45]:
Christmas_df=df[(df['playlist'] == "Christmas Hits") & (df['song_duration_insec'] >= 175) & (df['song_duration_insec'] <= 190)]

In [46]:
Christmas_df

playlist      song_name          singer  \
35  Christmas Hits    Sleigh Ride    The Ronettes   
39  Christmas Hits  Feliz Navidad  José Feliciano   
45  Christmas Hits      Mistletoe   Justin Bieber   

                                         album song_duration  \
35  A Christmas Gift For You From Phil Spector          3:01   
39                           The Definite Best          3:02   
45        Under The Mistletoe (Deluxe Edition)          3:02   

    song_duration_insec  
35                  181  
39                  182  
45                  182

In [47]:
len(Christmas_df)

3

In [48]:
Sleep_df=df[(df['playlist'] == "Sleep") & (df['song_duration_insec'] >= 175) & (df['song_duration_insec'] <= 190)]

In [49]:
Sleep_df

playlist   song_name                  singer              album  \
75    Sleep  Melancholy          Jonas Sandberg             Mosaic   
77    Sleep     Sakalas           Gabriel Derus            Sakalas   
78    Sleep   Spherical           Aiden Gaarder  Painted Illusions   
81    Sleep    Aesolace            Pollux Pluto           Aesolace   
85    Sleep  Retrospect                 Aveniós         Retrospect   
87    Sleep    Overcome  Anders-Petter Carlskog           Overcome   
89    Sleep       Epica             Ella Watson              Epica   
90    Sleep      Tender                 St/llar             Tender   
91    Sleep  Hemisphere           Solis Occasum         Hemisphere   
92    Sleep    Changing                De Janos           Changing   
94    Sleep   Following              Aquilonius          Following   
97    Sleep   Boundless                  Haldur          Boundless   

   song_duration  song_duration_insec  
75          2:59                  179  
77          2:56                  176  
78          2:55                  175  
81          3:01                  181  
85          3:04                  184  
87          3:10                  190  
89          3:09                  189  
90          3:03                  183  
91          3:02                  182  
92          3:03                  183  
94          3:05                  185  
97          2:58                  178

In [50]:
len(Sleep_df)

12

In [51]:
Today_df=df[(df['playlist'] == "Today’s Top Hits") & (df['song_duration_insec'] >= 175) & (df['song_duration_insec'] <= 190)]

In [52]:
Today_df

playlist                                song_name  \
3   Today’s Top Hits                                  Houdini   
7   Today’s Top Hits                             Cruel Summer   
17  Today’s Top Hits  Dance The Night - From Barbie The Album   
20  Today’s Top Hits                                  Feather   
21  Today’s Top Hits      Seven (feat. Latto) (Explicit Ver.)   
22  Today’s Top Hits                             Stick Season   
23  Today’s Top Hits                              Si No Estás   

               singer                                    album song_duration  \
3            Dua Lipa                                  Houdini          3:05   
7        Taylor Swift                                    Lover          2:58   
17           Dua Lipa  Dance The Night (From Barbie The Album)          2:56   
20  Sabrina Carpenter                 emails i can’t send fwd:          3:05   
21          Jung Kook                      Seven (feat. Latto)          3:04   
22         Noah Kahan                             Stick Season          3:02   
23     iñigo quintero                              Si No Estás          3:04   

    song_duration_insec  
3                   185  
7                   178  
17                  176  
20                  185  
21                  184  
22                  182  
23                  184

In [53]:
len(Today_df)

7

Therefore, Sleep is the playlist which contain the highest songs with average duration of 183 seconds (ie, the average duration of songs of whole dataset)

## Q2.How many unique singers contribute to each playlist, and which playlist has the highest diversity of singers?

In [59]:
unique_playlist=df['playlist'].unique()

In [60]:
unique_playlist

array(['Today’s Top Hits', 'Christmas Hits', 'Calming Acoustic', 'Sleep'],
      dtype=object)

In [64]:
df2_1=df[(df['playlist'] == "Today’s Top Hits")]

In [65]:
df2_1

playlist                                          song_name  \
0   Today’s Top Hits                                              Water   
1   Today’s Top Hits                                        Lovin On Me   
2   Today’s Top Hits                                             greedy   
3   Today’s Top Hits                                            Houdini   
4   Today’s Top Hits                                           MY HOUSE   
5   Today’s Top Hits                                          Strangers   
6   Today’s Top Hits                                             Snooze   
7   Today’s Top Hits                                       Cruel Summer   
8   Today’s Top Hits                               Standing Next to You   
9   Today’s Top Hits                                             MONACO   
10  Today’s Top Hits                                 Paint The Town Red   
11  Today’s Top Hits                                            vampire   
12  Today’s Top Hits      I Remember Everything (feat. Kacey Musgraves)   
13  Today’s Top Hits                              My Love Mine All Mine   
14  Today’s Top Hits                      Calm Down (with Selena Gomez)   
15  Today’s Top Hits                  You’re Losing Me (From The Vault)   
16  Today’s Top Hits                                        Agora Hills   
17  Today’s Top Hits            Dance The Night - From Barbie The Album   
18  Today’s Top Hits  Popular (with Playboi Carti & Madonna) - The I...   
19  Today’s Top Hits                                               exes   
20  Today’s Top Hits                                            Feather   
21  Today’s Top Hits                Seven (feat. Latto) (Explicit Ver.)   
22  Today’s Top Hits                                       Stick Season   
23  Today’s Top Hits                                        Si No Estás   
24  Today’s Top Hits                                       Lose Control   

               singer                                              album  \
0                Tyla                                              Water   
1         Jack Harlow                                        Lovin On Me   
2          Tate McRae                                             greedy   
3            Dua Lipa                                            Houdini   
4             Beyoncé                                           MY HOUSE   
5         Kenya Grace                                          Strangers   
6                 SZA                                                SOS   
7        Taylor Swift                                              Lover   
8           Jung Kook                                             GOLDEN   
9           Bad Bunny                nadie sabe lo que va a pasar mañana   
10           Doja Cat                                 Paint The Town Red   
11     Olivia Rodrigo                                            vampire   
12         Zach Bryan                                         Zach Bryan   
13             Mitski             The Land Is Inhospitable and So Are We   
14               Rema                      Calm Down (with Selena Gomez)   
15       Taylor Swift                  You're Losing Me (From The Vault)   
16           Doja Cat                                            Scarlet   
17           Dua Lipa            Dance The Night (From Barbie The Album)   
18         The Weeknd  Popular [The Idol Vol. 1 (Music from the HBO O...   
19         Tate McRae                                               exes   
20  Sabrina Carpenter                           emails i can’t send fwd:   
21          Jung Kook                                Seven (feat. Latto)   
22         Noah Kahan                                       Stick Season   
23     iñigo quintero                                        Si No Estás   
24        Teddy Swims         I've Tried Everything But Therapy (Part 1)   

   song_duration  song_duration_insec  
0           3:20                  200  
1   

In [68]:
len(df2_1['singer'].unique())

20

In [101]:
max=-1
final_playlist=""
for playlist in unique_playlist:
    df2=df[(df['playlist'] == playlist)]
    length=len(df2['singer'].unique())
    print(f"{length} number of unique singer contribute the {playlist} ")
    if length>max:
        max=length
        final_playlist=playlist

print(f"The {final_playlist}  playlist contains songs that refer to the most unique singers, which is equal to  {length}.")


20 number of unique singer contribute the Today’s Top Hits 
23 number of unique singer contribute the Christmas Hits 
23 number of unique singer contribute the Calming Acoustic 
24 number of unique singer contribute the Sleep 
The Sleep  playlist contains songs that refer to the most unique singers, which is equal to  24.


And Hence , Sleep playlist contain the songs of high diversity of singers

## Q3.Identify the top album/albums with the most songs in playlists, and for each of these albums, what is the total duration of their songs across all playlists?

In [85]:
df3=df.groupby('album').count()

In [86]:
df3

playlist  song_name  singer  \
album                                                                     
A Christmas Gift For You From Phil Spector         1          1       1   
Act of Kindness                                    1          1       1   
Aesolace                                           1          1       1   
Aligned Adventures                                 1          1       1   
Aquamarine                                         1          1       1   
...                                              ...        ...     ...   
emails i can’t send fwd:                           1          1       1   
exes                                               1          1       1   
greedy                                             1          1       1   
nadie sabe lo que va a pasar mañana                1          1       1   
vampire                                            1          1       1   

                                            song_duration  song_duration_insec  
album                                                                           
A Christmas Gift For You From Phil Spector              1                    1  
Act of Kindness                                         1                    1  
Aesolace                                                1                    1  
Aligned Adventures                                      1                    1  
Aquamarine                                              1                    1  
...                                                   ...                  ...  
emails i can’t send fwd:                                1                    1  
exes                                                    1                    1  
greedy                                                  1                    1  
nadie sabe lo que va a pasar mañana                     1                    1  
vampire                                                 1                    1  

[97 rows x 5 columns]

In [91]:
df3.sort_values(by=['playlist'],ascending=False)

playlist  song_name  singer  \
album                                                                     
Christmas (Deluxe Special Edition)                 2          2       2   
A Christmas Gift For You From Phil Spector         1          1       1   
Portrait                                           1          1       1   
Si No Estás                                        1          1       1   
Seven (feat. Latto)                                1          1       1   
...                                              ...        ...     ...   
I've Tried Everything But Therapy (Part 1)         1          1       1   
Houdini                                            1          1       1   
Herinneringen                                      1          1       1   
Hemisphere                                         1          1       1   
vampire                                            1          1       1   

                                            song_duration  song_duration_insec  
album                                                                           
Christmas (Deluxe Special Edition)                      2                    2  
A Christmas Gift For You From Phil Spector              1                    1  
Portrait                                                1                    1  
Si No Estás                                             1                    1  
Seven (feat. Latto)                                     1                    1  
...                                                   ...                  ...  
I've Tried Everything But Therapy (Part 1)              1                    1  
Houdini                                                 1                    1  
Herinneringen                                           1                    1  
Hemisphere                                              1                    1  
vampire                                                 1                    1  

[97 rows x 5 columns]

We have only one album which have more number of songs then other album in playlist

In [95]:
df3_2=df[df['album']=='Christmas (Deluxe Special Edition)']

In [96]:
df3_2

playlist                                    song_name  \
29  Christmas Hits  It's Beginning to Look a Lot like Christmas   
38  Christmas Hits                        Holly Jolly Christmas   

           singer                               album song_duration  \
29  Michael Bublé  Christmas (Deluxe Special Edition)          3:26   
38  Michael Bublé  Christmas (Deluxe Special Edition)          1:59   

    song_duration_insec  
29                  206  
38                  119

In [99]:
sum=0
for i in df3_2['song_duration_insec']:
    sum=sum+i
print(f"Total songs duration in sec for album Christmas (Deluxe Special Edition) is {sum}")

Total songs duration in sec for album Christmas (Deluxe Special Edition) is 325


Hence, We can conclude that "Christmas (Deluxe Special Edition)" is the album with the highest number of songs in playlists with total song duration of songs in sec is 325 

### Conclusion

### Q1.What is the average duration of songs in each playlist, and which playlist has the highest average song duration?

| Playlist           | Song Duration (seconds) |
|---------------------|-------------------------|
| Calming Acoustic    | 163.083333              |
| Christmas Hits      | 183.480000              |
| Sleep               | 183.125000              |
| Today’s Top Hits    | 201.200000              |


Therefore, Sleep is the playlist which contain the highest number of songs with average duration of 183 seconds (ie, the average duration of songs of whole dataset)

### Q2.How many unique singers contribute to each playlist, and which playlist has the highest diversity of singers?

- 20 number of unique singer contribute the Today’s Top Hits 
- 23 number of unique singer contribute the Christmas Hits 
- 23 number of unique singer contribute the Calming Acoustic 
- 24 number of unique singer contribute the Sleep 

Hence, Sleep playlist contain the songs of high diversity of singers

### Q3.Identify the top album or albums with the most songs in playlists, and for each of these albums, what is the total duration of their songs across all playlists?

Hence, We can conclude that "Christmas (Deluxe Special Edition)" is the album with the highest number of songs in playlists with total song duration in sec is 325

# End